File: backend/app/modules/billing/cards.py
"""Reading a card number, and forgetting it.

Pure functions, in their own module and importing nothing from the rest of the app, so
the one piece of code that ever holds a full card number is small enough to read in one
sitting and testable without a database.

**Everything here takes a number and returns something that is not one.** There is no
function that stores, echoes, or logs the digits it was given, and
:func:`inspect_card_number` returns exactly the two facts that get persisted - the
network and the last four. That is the whole contract: the PAN exists as a local
variable for the length of one call and then it is gone.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Final

from app.modules.billing.models import CardNetwork

#: Separators people type or paste. Stripped before anything else looks at the value.
_SEPARATORS: Final = re.compile(r"[\s\-]")

#: The range real card numbers fall in. ISO/IEC 7812 allows up to 19 digits; the
#: shortest scheme still in circulation is 12 (some Maestro).
MIN_DIGITS: Final = 12
MAX_DIGITS: Final = 19


@dataclass(frozen=True, slots=True)
class CardIdentity:
    """What is safe to keep about a card.

    Deliberately does *not* carry the number it came from - there is nowhere for a
    caller to accidentally read it back out of. ``checksum_ok`` is a verdict *about* the
    number, not a piece of it.
    """

    network: CardNetwork
    last4: str

    #: Whether the number passed its Luhn check digit.
    #:
    #: **A card that fails this is not a valid card, and the service refuses it.** Carried as
    #: a field rather than folded into a `None` return so the caller can tell the two kinds
    #: of failure apart and say which happened: "that is the wrong number of digits" and
    #: "those digits do not check out" are different mistakes with different fixes, and one
    #: generic message for both is the kind of thing that leaves someone re-typing a correct
    #: number.
    checksum_ok: bool


def normalise_card_number(raw: str) -> str:
    """Strip spaces and dashes. Returns digits, or whatever else was in there."""
    return _SEPARATORS.sub("", raw.strip())


def is_plausible_card_number(digits: str) -> bool:
    """Length and character check, before the arithmetic.

    Separate from the Luhn test so a caller can tell "that is not a card number" from
    "that is a card number with a typo in it", which are different messages to show.
    """
    return digits.isdigit() and MIN_DIGITS <= len(digits) <= MAX_DIGITS


def passes_luhn(digits: str) -> bool:
    """The check digit, as every issuer computes it.

    Worth doing rather than accepting any 16 digits: Luhn catches every single-digit
    typo and almost every transposition, which is the entire class of mistake someone
    makes copying a number off a card. It says nothing about whether the card exists -
    that is not knowable here and not needed, because nothing is charged.
    """
    if not digits.isdigit():
        return False

    total = 0
    # Doubling every second digit from the right, so the parity depends on the length.
    parity = len(digits) % 2
    for index, character in enumerate(digits):
        value = int(character)
        if index % 2 == parity:
            value *= 2
            if value > 9:
                value -= 9
        total += value
    return total % 10 == 0


#: Leading-digit ranges, longest prefix first so a more specific rule wins.
#:
#: **RuPay is here and near the front on purpose.** This product is built for Indian
#: small businesses, where RuPay is the most common domestic scheme - and several of its
#: ranges (``60``, ``65``, ``81``) overlap Discover and Maestro, so ordering decides the
#: answer. Getting it wrong is cosmetic rather than dangerous, but showing an Indian
#: shopkeeper "Discover" for their RuPay card reads as software that does not know where
#: it is.
_PREFIXES: Final[tuple[tuple[tuple[str, ...], CardNetwork], ...]] = (
    (("34", "37"), CardNetwork.AMEX),
    (("6521", "6522", "60", "81", "82", "508"), CardNetwork.RUPAY),
    (("4",), CardNetwork.VISA),
    (
        (
            "51",
            "52",
            "53",
            "54",
            "55",
            "2221",
            "2222",
            "2223",
            "2224",
            "2225",
            "2226",
            "2227",
            "2228",
            "2229",
            "223",
            "224",
            "225",
            "226",
            "227",
            "228",
            "229",
            "23",
            "24",
            "25",
            "26",
            "270",
            "271",
            "2720",
        ),
        CardNetwork.MASTERCARD,
    ),
    (("6011", "644", "645", "646", "647", "648", "649", "65"), CardNetwork.DISCOVER),
    (("300", "301", "302", "303", "304", "305", "3095", "36", "38", "39"), CardNetwork.DINERS),
    (
        ("3528", "3529", "353", "354", "355", "356", "357", "358"),
        CardNetwork.JCB,
    ),
    (("50", "56", "57", "58", "639", "67"), CardNetwork.MAESTRO),
)


def detect_network(digits: str) -> CardNetwork:
    """Which scheme a number belongs to, from its leading digits.

    Matched on the longest prefix so ``6521`` reads as RuPay rather than as Discover's
    ``65``. Anything unrecognised is :attr:`CardNetwork.OTHER` - the card is still
    perfectly usable, the software simply does not claim to know the scheme.
    """
    best: CardNetwork = CardNetwork.OTHER
    best_length = 0
    for prefixes, network in _PREFIXES:
        for prefix in prefixes:
            if digits.startswith(prefix) and len(prefix) > best_length:
                best = network
                best_length = len(prefix)
    return best


def inspect_card_number(raw: str) -> CardIdentity | None:
    """Everything the system keeps about a card, or ``None`` if it is not one.

    The single entry point, so there is exactly one place a full number is handled.
    Returning ``None`` rather than raising keeps this module free of the app's exception
    hierarchy - the caller owns the message, because the caller knows whether it is
    answering a form field or a script.

    **``None`` means the wrong shape.** A failed check digit comes back on
    :attr:`CardIdentity.checksum_ok` instead - both are refused by the caller, but they are
    reported apart so the message can say which one happened. "Twelve to nineteen digits" and
    "those digits do not check out" are different mistakes with different fixes.
    """
    digits = normalise_card_number(raw)
    if not is_plausible_card_number(digits):
        return None
    return CardIdentity(
        network=detect_network(digits),
        last4=digits[-4:],
        checksum_ok=passes_luhn(digits),
    )

File: frontend/src/features/billing/cards.ts
/**
 * Card presentation and the one check worth doing client-side.
 *
 * A plain module, not a component file: shared by the panel on Billing and by the Accounts
 * page, and a file that exports both components and helpers breaks fast refresh for
 * everything in it.
 */
import type { CardNetwork } from '@/features/billing/api';

export const NETWORK_LABELS: Record<CardNetwork, string> = {
  visa: 'Visa',
  mastercard: 'Mastercard',
  rupay: 'RuPay',
  amex: 'Amex',
  discover: 'Discover',
  diners: 'Diners Club',
  jcb: 'JCB',
  maestro: 'Maestro',
  // Not "Unknown": the card works perfectly well, the software just does not claim to
  // recognise the scheme from its leading digits.
  other: 'Card',
};

/** The range real card numbers fall in - ISO/IEC 7812, and the shortest Maestro. */
export const MIN_DIGITS = 12;
export const MAX_DIGITS = 19;

/** Strip the separators a person types or pastes. */
export function normaliseCardNumber(raw: string): string {
  return raw.replace(/[\s-]/g, '');
}

/**
 * Length and character check, kept separate from {@link passesLuhn}.
 *
 * The split lets a form tell "that is not a card number" from "that is a card number with a
 * typo in it", which are different things to say to someone - and stops it complaining
 * about a check digit before they have finished typing.
 */
export function isPlausibleCardNumber(digits: string): boolean {
  return digits.length >= MIN_DIGITS && digits.length <= MAX_DIGITS && /^\d+$/.test(digits);
}

/**
 * Why the card number is not acceptable yet, or `null` when it is.
 *
 * **This exists because the "Add card" button used to disable itself in silence.** Type
 * eleven digits and nothing happened: no message, no error, a greyed-out button, and no way
 * to tell that one more digit was needed. The rules were only ever expressed as a boolean
 * for the button's `disabled` prop, which is not something a person can read.
 *
 * Returns a message for every state that blocks saving, and counts digits out loud while
 * the number is too short - "11 of at least 12 digits" answers the question the greyed-out
 * button raises, where "Invalid" would not.
 */
export function cardNumberProblem(digits: string): string | null {
  if (digits === '') return null;
  if (!/^\d+$/.test(digits)) return 'Digits only.';
  if (digits.length < MIN_DIGITS) {
    // Counted out loud while the number is short, rather than "invalid": one digit missing
    // and one digit wrong need different things from the user, and until the length is right
    // there is nothing meaningful to say about the digits themselves.
    return `${digits.length} of at least ${MIN_DIGITS} digits.`;
  }
  if (digits.length > MAX_DIGITS) {
    return `A card number is at most ${MAX_DIGITS} digits.`;
  }
  // The Luhn check digit, and it **blocks the save**. Every real card number satisfies it,
  // so failing means a typo - and the last four digits are how this card gets recognised on
  // an entry months later, so a wrong label defeats the point of keeping one. Only asked
  // once the length is plausible, or it would fire on every keystroke while typing.
  if (!passesLuhn(digits)) return 'That is not a valid card number - check the digits.';
  return null;
}

/**
 * The Luhn check digit, so a typo is caught before a round trip.
 *
 * A duplicate of the server's check, deliberately: the server stays the authority, this
 * only saves someone a request to be told they mistyped one digit. Safe to duplicate
 * because Luhn is a fixed algorithm that cannot drift - unlike the issuer-range table that
 * names the scheme, which is left entirely to the server.
 */
export function passesLuhn(digits: string): boolean {
  if (!/^\d+$/.test(digits)) return false;

  let total = 0;
  // Doubling every second digit from the right, so the parity depends on the length.
  const parity = digits.length % 2;
  for (let index = 0; index < digits.length; index += 1) {
    let value = Number(digits[index]);
    if (index % 2 === parity) {
      value *= 2;
      if (value > 9) value -= 9;
    }
    total += value;
  }
  return total % 10 === 0;
}
